Merge IConfigurationRoot into IConfiguration and use new helper ChangeToken

HaoK opened the pull request:
Depends on https://github.com/aspnet/Common/pull/63

cc @divega

divega commented on src/Microsoft.Extensions.Configuration.Abstractions/IConfiguration.cs:
```diff
@@ -36,5 +36,6 @@ public interface IConfiguration
         IEnumerable<IConfigurationSection> GetChildren();
 
         IChangeToken GetReloadToken();
+        IConfigurationRoot Root { get; }
```
With this change IConfigurationRoot will also have a Root property, so with the default application template Intellisense on `Configuration` will show an almost meaningless property to millions of customers :smile:

Can we add the `Root` property only on the `IConfigurationSection` interface instead? We can always define the `GetRoot(IConfiguration)` extension method to do the right thing and put it in a namespace that isn't in the default template so that you opt in to see it?

HaoK replied:
I guess I still think of these as a node interface.  So I don't think its that bad, since the root of a tree, either has a null parent, or itself.  Am I mistaken to think of IConfiguration in this way?  I still translate it to IConfigurationNode, of which IConfigurationRoot happens to be the root.

HaoK replied:
Minor thing while I'm here, should we make GetReloadToken() into a read only property instead of a method?  Its just returning a field currently...

HaoK replied:
Alternatively we can just nuke it entirely since we are exposing root now, and it was just directing to the root method, we can remove this off IConfiguration and leave it only on IConfigurationRoot next to Reload()...even better?

rynowak replied:
I feel like `GetReloadToken()` as a method is better because you'll always been accessing the token in a 'loop'. You never really want to cache the result.

divega commented:
cc @rynowak @lodejard

HaoK commented:
Updated with new tests verifying ChangeToken Dispose behavior.  Still need to resolve IConfigurationRoot, @divega are you still strongly opposed to just having it on IConfiguration?

HaoK commented on test/Microsoft.Extensions.Configuration.FileProviderExtensions.Test/ConfigurationRootExtensionsTest.cs:
```diff
@@ -40,12 +40,59 @@ public void ReloadOnChanged_GetTokenBeforeReload()
             Assert.Equal(1, configuration.ReloadCount);
         }
 
+        [Fact]
```
@rynowak Please confirm that the following 3 behavior verification tests align with your expectations :)

rynowak commented:
:shipit: from me. Tests look right. I'll let you and @divega resolve the `IConfigurationRoot` discussion

HaoK commented:
@divega are you happy with the current changes for initial checkin?  We can always tweak this later

lodejard commented:
Why do we need Root property?

HaoK commented:
@lodejard It was needed to expose an overload of the configure that binds options to a particular `IConfiguration`: which needs to call Reload.   Given that GetReloadToken basically is directly forwarding to the root and is on IConfiguration, I still argue that we should rename Reload => ReloadAll and have it on IConfiguration as well, and then we don't need Root and or even IConfigurationRoot as an interface at all (yes I've had it in for that guy from the beginning :)

```
    public class ConfigurationChangeTokenSource<TOptions> : IOptionsChangeTokenSource<TOptions>
    {
        private IConfiguration _config;

        public ConfigurationChangeTokenSource(IConfiguration config);
        public IChangeToken GetChangeToken()
        {
            return _config.GetReloadToken();
        }

        public void Changed()
        {
            _config.Root.Reload();
        }
    }

HaoK commented:
Updated: merged IConfigurationRoot back into IConfiguration and renamed Reload -> ReloadAll.

@divega not sure we need the All suffix, do you think its enough to put it in the doc comments that Reload affects the entire configuration and dropping the All?

HaoK commented:
This is stale
